Add onPan and onZoom to the PanZoom interaction

SamMorrowDrums opened the pull request:
As mentioned in #3086 I am proposing to add events for Pan and Zoom, so that users can automatically retrieve more data for charts.

The proposed API is such:

``` javascript
panZoomInteraction = new Plottable.Interactions.PanZoom();

panZoomInteraction.onPan(callback);
panZoomInteraction.offPan(callback);

panZoomInteraction.onZoom(callback);
panZoomInteraction.offZoom(callback);

```

Callbacks occur on the most applicable end event (touchend, dragend or scroll), in an attempt to minimise the number of calls where possible, as this will most likely be used for retrieving server data.

SamMorrowDrums committed:
Add onPan and onZoom to the PanZoom interaction

adidahiya commented:
Nice, this looks clean & well-tested. Is it easy to tell what new data you should retrieve from a server on these events? What might that code look like for Plottable users?

adidahiya commented:
Actually, can you change the API to `onPanEnd` and `onZoomEnd`? Looking at https://github.com/palantir/plottable/issues/2651, it seems like we'd also like to have hooks into the _start_ of these interactions. Naming the ones in this PR more precisely will make it easier to add those hooks later.

// cc @jtlan for thoughts

SamMorrowDrums commented:
Yes, that makes sense. I'll update the naming and push again.

I'm actually working on code right now that would fire on this event as we speak! Naturally there are a lot of race conditions that can occur getting async data. Also, there is no actual scroll-end event, so a lot of people implement a timer, and don't fire the true callback until they stop upping the timer.

I'm not sure that the timer code belongs in Plottable codebase though. For now I'll just update the names, and work on implementation of the calling code!

SamMorrowDrums committed:
Update API to panEnd and zoomEnd events

SamMorrowDrums commented:
Also, for reference I'm using an intermediate callback that gets the x-axis domain for my data gathering:

``` javascript
panZoomInteraction.onZoomEnd(function () {
        const d = that.xScale.domain();
        callback(d[0], d[1]);
});
```

jtlan commented:
> Is it easy to tell what new data you should retrieve from a server on these events? What might that code look like for Plottable users?

I would expect that users would invoke `domain()` calls to figure out the view window at that time. Alternatively the callbacks could return the domain of the `Scale`s at that time, but there wouldn't be any type information because `PanZoom` interaction isn't genericized.

jtlan commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -424,5 +437,49 @@ namespace Plottable.Interactions {
+     * Adds a callback to be called when panning ends.
+     *
+     * @param {PanCallback} callback
+     * @returns {Event} The calling PanZoom Interaction.
```
This and other registration calls should return `{this}` not `{Event}`.

jtlan commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -1,4 +1,8 @@
 namespace Plottable.Interactions {
+
+  export type PanCallback = (e: Event) => void;
+  export type ZoomCallback = (e: Event) => void;
```
Not sure these callbacks should take `Event`s -- that's sort of atypical if you compare to the callbacks passed to `Click`, `Drag`, and `Pointer` `Interaction`s. I would expect the final domains of the `Scale`s or something like that.

SamMorrowDrums replied:
Yeah, I was a little unsure as to what to return, but you can take multiple
scales right, so would it be an array of domains returned?

On Thu, 14 Jul 2016, 19:27 jtlan, notifications@github.com wrote:

> In src/interactions/panZoomInteraction.ts
> https://github.com/palantir/plottable/pull/3087#discussion_r70857970:
> 
> > @@ -1,4 +1,8 @@
> >  namespace Plottable.Interactions {
> > +
> > -  export type PanCallback = (e: Event) => void;
> > -  export type ZoomCallback = (e: Event) => void;
> 
> Not sure these callbacks should take Events -- that's sort of atypical if
> you compare to the callbacks passed to Click, Drag, and Pointer
> Interactions. I would expect the final domains of the scales or something
> like that.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/palantir/plottable/pull/3087/files/cf951aba815a4c8d2526d68def596f8869c1ed41#r70857970,
> or mute the thread
> https://github.com/notifications/unsubscribe/AElqXol0-UeMq4VU9_qldkW-AYz-FEfmks5qVn-bgaJpZM4JMMcF
> .

jtlan replied:
Ideally it would be something like:

```
{
  xDomains: Map<Scale, X[]>;
  yDomains: Map<Scale, Y[]>;
}
```

but the typing on `PanZoom` doesn't allow for that since it's not generic on `<X, Y>` (also object-keyed `Map` isn't standard yet). For now we might want the callback to pass nothing (but still keep it type-defined), since the users can retrieve the domain information from the `Scale`s directly.

SamMorrowDrums committed:
Update typing and documentation

SamMorrowDrums commented:
OK @adidahiya & @jtlan I've fixed the documentation and removed the event arguments from the types, as it sounds like we are not yet decided on best approach for passing the domains (although I agree that is certainly major use case, and users will likely do this manually for now)

I think, if we want a more realistic Zoom event from scroll wheel, we need to implement an intermediate callback that sets a timer (of say 100ms), and keeps cancelling timer and extending until no wheel events occur in time, and callback fires.

Should this work on the event timer be completed now, or are we better progressing with the events, and then building on them (as part of the Zoom and Pan start event work)?

SamMorrowDrums commented:
```

SamMorrowDrums commented:
I'm doing the timeout in my own code, and its behaviour is pretty solid. Here is roughly what I'm doing in my own codebase:

``` javascript
import Plottable from "plottable";

class PanZoom extends Plottable.Interactions.PanZoom {
    constructor (cb) {
        super();
        this.timer = null;
        this.onDateChange = cb;
        if (cb) {
            this._dragInteraction.onDragEnd(cb);
        }
    }

    _handleWheelEvent (p, e) {
        const that = this;
        super._handleWheelEvent(p, e);
        if (this.onDateChange) {
            if (this.timer) {
                clearTimeout(this.timer);
            }
            this.timer = setTimeout(function () {
                that.onDateChange();
            }, 100);
        }
    }

    _handleTouchEnd (ids, idToPoint, e) {
        super._handleTouchEnd(ids, idToPoint, e);
        if (this.onDateChange && this._touchIds.size() > 0) {
            this.onDateChange();
        }
    }
}

export default PanZoom;
```

SamMorrowDrums commented:
@adidahiya you asked how I was planning to implement the calling code, current process is:
- Callback fires
- Pass changed domain into an array
- Check if it's within existing data
- Send data request (and empty array)
- Wait until data gathered
- If more callbacks have fired during data gathering, get the min and max of the domains from callbacks, and then if more data required data, send request (clear array again) and repeat.

SamMorrowDrums commented:
I have it working perfectly for me, with the data gathering. Really nice feel to scroll eternally through time series and zoom in and out, without any restriction. Obviously, there are some limits, for example you might constrain the array length, and start to pull out data that's no longer visible, but I'm very happy with this behaviour. It really lifts the experience!

SamMorrowDrums commented:
@adidahiya is there anything further I should do to finish this?

adidahiya commented:
@SamMorrowDrums sorry for the delay, this looks great!

A5048 commented:
> As mentioned in #3086 I am proposing to add events for Pan and Zoom, so that users can automatically retrieve more data for charts.
> 
> The proposed API is such:
> 
> ``` javascript
> panZoomInteraction = new Plottable.Interactions.PanZoom();
> 
> panZoomInteraction.onPan(callback);
> panZoomInteraction.offPan(callback);
> 
> panZoomInteraction.onZoom(callback);
> panZoomInteraction.offZoom(callback);
> 
> ```
> 
> Callbacks occur on the most applicable end event (touchend, dragend or scroll), in an attempt to minimise the number of calls where possible, as this will most likely be used for retrieving server data.
>

A5048 reviewed (commented):
@arash_sharifi_basht
